gr-azure-software-radio/python/eventhub_source.py

```diff
@@ -0,0 +1,147 @@
+    supported by the DefaultAzureCredential class, such as environment variables.
+
+        Args:
+    Auth Method: Determines how to authenticate to Event Hub. May be one of
+        "connection_string", "sas", or "default".
+    Connection String: Event Hub connection string used for
+        authentication if Auth Method is "connection_string".
+    Shared Access Signature (SAS) token. This is required if using "default" or
+        "sas" authentication.
+    Event Hub Host Name:  The fully qualified host name for the Event Hub namespace. This is
+        required if using "default" or "sas" authentication.
+    Event Hub Name: The Event Hub must be created in Azure before trying to
+        receive events from.
+    Consumer Group: The consumer group to receive events from Event Hub.
+    Partition ID: The partition ID to receive events from.
+    Starting Position: The position of an event in the Event Hub partition.
+    """
+    # pylint:  disable=too-many-arguments, no-member
+    def __init__(
+            self,
+            authentication_method: str = "default",
+            connection_str: str = None,
+            sas_token: str = None,
+            eventhub_host_name: str = None,
+            eventhub_name: str = None,
+            consumer_group: str = None,
+            partition_id: str = None,
+            starting_position=None):
+
+        gr.sync_block.__init__(self,
+                               name="eventhub_source",
+                               in_sig=[],
+                               out_sig=[])
+        self.starting_position = starting_position
+        self.partition_id = partition_id
+
+        self.eventhub_consumer = get_eventhub_consumer_client(
+            authentication_method=authentication_method,
+            connection_str=connection_str,
+            sas_token=sas_token,
+            eventhub_host_name=eventhub_host_name,
+            eventhub_name=eventhub_name,
+            consumer_group=consumer_group,
+        )
+        
+        self.message_port_register_out(pmt.intern('out'))
+
+        self.rec_thread = threading.Thread(target=self.receive)
+        self.rec_thread.start()
+
+    def receive(self):
+        """
+        Receive events from event hub given the specified partition and starting position.
+        The receive call is blocking and must be run in a dedicated thread.
+        """
+        self.eventhub_consumer.receive(
+            on_event=self.on_event,
+            partition_id=self.partition_id,
+            starting_position=self.starting_position)
+
+    def on_event(self, _partition_context, event):
+        """
+        Convert the received JSON message to PMT. The expected Event Hub event only has one message batched.
+        """
+        # pylint: disable=no-member
+        msg = json.loads(list(event.body)[0])
+        pmsg = pmt.to_pmt(msg)
+        self.message_port_pub(pmt.intern("out"), pmsg)
+
+    def stop(self):
+        self.eventhub_consumer.close()
```

Comment:
I think we may be fine since @jo-rivera in not using multiprocess threading, so should die with the this process. 

we should confirm though @jo-rivera

Reply:
it does but is good practice to stop the thread. I updated the code.

Reply:
Ah and the eventhub_comsumer.close() shuts down the self.eventhub_consumer.receive() call. This is tidy, I like it. 

I never knew self.eventhub_consumer.receive would keep running across multiple batches of messages, I always thought it would return after it got to the end of the messages that were already on the partition. This is cool, nicely done.